### ISSUE TITLE
brackets [] not auto-closing

### ISSUE THREAD

**Issue by andmkent** [greghendershott/racket-mode#155]
When I type an opening paren '(' or brace '{' the matching closing paren/bracket is automatically inserted.

When I type an opening bracket '[' however, I do not get the matching ']'.

I'm not using any fancy paren-related features (e.g. paredit), I double checked that the standard built in eletric-pair-mode is on, and I looked in racket-common.el and it looks like there are the correct entries for brackets:

```lisp
    (modify-syntax-entry ?\[ "(]  " st)
    (modify-syntax-entry ?\] ")[  " st)
```

So... I'm not sure why I'm seeing this behavior. Any thoughts?

What's interesting is that, in an elisp file, I get the correct matching behavior for all 3 delimiters, but in racket-mode just for parens and brackets.

**Comment by andmkent** [comment]
Update: so, it seems it's related to this line in racket-mode.el:

```lisp
     ("["           racket-smart-open-bracket)
```

Commenting that out fixes the problem (although it probably breaks the ability to use the smart-open-bracket feature, obviously)

**Comment by greghendershott** [comment]
Thanks for the report!

I've never used `electric-pair-mode`. Now that I try it, I can repro what you describe.

I had added code to check for `paredit-mode`, in which case I call `paredit-open-square` or `paredit-open-round` instead of `insert`-ing the character directly.

Probably I need to do something similar, for `electric-pair-mode`. Or, perhaps I'm barking up the wrong tree here. I wanted to make a simple on/off switch for people. However it would probably be more Emacs-y to make people bind <kbd>[</kbd> to `racket-smart-open-bracket` IFF they want the smart open bracket stuff.

Which, by the way, should be another non-Elisp-modifying way to do this, I think?  In your init.el, bind bind <kbd>[</kbd> to `self-insert-command` in `racket-mode-map`?

[Which, in fact, might be what I should do instead of calling `insert`, to be compatible with `paredit`, `electric-pair-mode`, `smartparens`, and who knows what else is out there....]

TL;DR thanks again for the report, and I'll figure this out!

**Comment by andmkent** [comment]
Just a follow-up/FYI: It seems the fix for this *does* allow '[' to autocomplete and insert the ']' to close the brackets when using electric-pair-mode, but if you do this with some text selected, it obliterates the text, and does not insert the brackets on either side of the selection (which is what electric-pair-mode and DrRacket will do).

Fortunately if you don't use the racket-smart-open-bracket feature, the commenting out the line approach (I mentioned before) still works =)

**Comment by greghendershott** [comment]
Hmm. That seems to be a direct result of fixing #159 as requested by @cpitclaudel. The motivation for that was, if `electric-pair-mode` is _not_ enabled and `delete-selection-mode` _is_ enabled, then you would reasonably expect typing <kbd>[</kbd> to replace the selection.

When I `(put 'racket-smart-open-bracket 'delete-selection nil)`, it works as you want with `electric-pair-mode`.

Pondering how to resolve this dilemma....

**Comment by greghendershott** [comment]
I could change the whole "smart bracket open" feature into a minor-mode, and have it bind <kbd>[</kbd> -- or not.  That way the key wouldn't be bound unless people enable such a `racket-smart-bracket-open-mode`, and this issue would be less likely to occur.

But if people do want both `racket-smart-bracket-open-mode` and `electric-pair-mode`, the problem still exists for those people.

---

I guess I could test explicitly for `electric-pair-mode`, but that feels hacky. Also it's not immediately obvious to me how I would set the `'delete-section` property to be "dynamic" as opposed to `t` or `nil`. Maybe it can be a function? I'll check that....

---

FWIW this is N/A for `paredit-mode`, because it rebinds all the open and close keys to its own commands.  However `electric-pair-mode` doesn't rebind keys to commands, it works IIRC via hooks.

**Comment by greghendershott** [comment]
> Also it's not immediately obvious to me how I would set the 'delete-section property to be "dynamic" as opposed to t or nil. Maybe it can be a function? I'll check that....

Fortunately the `'delete-selection` property value _can_ be a function. I'll try that.

**Comment by greghendershott** [comment]
Although I need to think and test, more, this seems to work so far:

```
;; Set property for delete-selection-mode. Inserted text should
;; replace the selection unless electric-pair-mode is enabled.
(put 'racket-smart-open-bracket 'delete-selection
     (lambda ()
       (not (memq 'electric-pair-mode minor-mode-list))))
```

---

Looking at some other code related to paren-like things, I see some opportunities to simplify. Also I think I might be able to greatly simplify the paredit compatibility code.

**Comment by greghendershott** [comment]
@andmkent I pushed a commit with essentially that change.
